src/Avalonia.Input/KeyboardDevice.cs

```diff
@@ -122,18 +124,32 @@ public void ProcessRawEvent(RawInputEventArgs e)
 
                 if (text != null)
                 {
-                    var ev = new TextInputEventArgs()
+                    string cleanedText = RemoveInvalidCharacters(text.Text);
+                    if (!string.IsNullOrEmpty(cleanedText))
                     {
-                        Device = this,
-                        Text = text.Text,
-                        Source = element,
-                        RoutedEvent = InputElement.TextInputEvent
-                    };
-
-                    element.RaiseEvent(ev);
-                    e.Handled = ev.Handled;
+                        var ev = new TextInputEventArgs()
+                                 {
+                                     Device = this,
+                                     Text = cleanedText,
+                                     Source = element,
+                                     RoutedEvent = InputElement.TextInputEvent
+                                 };
+
+                        element.RaiseEvent(ev);
+                        e.Handled = ev.Handled;
+                    }
                 }
             }
         }
+
+        public string RemoveInvalidCharacters(string text)
+        {
+            for (var i = 0; i < invalidCharacters.Length; i++)
```

Comment:
Code is bloody inefficient, but I don't know if we care...

Reply:
Feel free to suggest a better way